### PR TITLE
Service Disposal Pattern

### PR DESCRIPTION
**Pull request by paulhazen**
This PR implements the disposal pattern for the `EOSService` base class and implementing classes, to guarantee cleanup tasks happen when and where appropriate.

#EOS-2032

### CODE REVIEW

File: Assets/Scripts/StandardSamples/Services/StatsService.cs [hunk 1]
```diff
@@ -64,13 +63,51 @@ public static StatsService Instance
...
+                // NOTE: Dispose of managed resources here.
+                Reset();
+            }
+
+            // NOTE: Free unmanaged resources here, and set large fields to null.
+
+            _disposed = true;
+
+            base.Dispose(disposing);
+        }
```
Comment (paulhazen) [inline Assets/Scripts/StandardSamples/Services/StatsService.cs]: This method is redundant - it should be removed from this class.

File: Assets/Scripts/StandardSamples/Services/EOSService.cs [hunk 2]
```diff
@@ -67,6 +72,63 @@ protected EOSService(bool requiresAuthentication = true)
...
+        /// end of that implementation. Implementing classes probably _should_
+        /// override this function.
+        /// </summary>
+        protected virtual void Reset()
+        {
+            // Set the updated event to null, which removes any subscribers.
+            Updated = null;
+
+            // Unsubscribe from the authentication changed event
+            AuthenticationListener.Instance.AuthenticationChanged -= OnAuthenticationChanged;
```
Comment (paulhazen) [inline Assets/Scripts/StandardSamples/Services/EOSService.cs]: This should be removed.

File: Assets/Scripts/StandardSamples/Services/EOSService.cs [hunk 3]
```diff
@@ -67,6 +72,63 @@ protected EOSService(bool requiresAuthentication = true)
...
+            }
+
+            if (disposing)
+            {
+                // NOTE: Dispose of managed resources here.
+                Reset();
+            }
+
+            // NOTE: Free unmanaged resources here, and set large fields to null.
+
```
Comment (paulhazen) [inline Assets/Scripts/StandardSamples/Services/EOSService.cs:113]: This is where the event for `AuthenticationListener` should be unsubscribed from.

**Review by arthur740212 — APPROVED** [review]
Only thing that confused me at first is the disposing and disposed.
In `~StatsService()` the Dispose(false) made me thought it was not going to do anything.
And AchievementService didn't have a destructor.

**Review by andrew-hirata-playeveryware — APPROVED** [review]

**Pull request merged**